src/libponyrt/sched/scheduler.c

```diff
@@ -333,16 +394,64 @@ static pony_actor_t* steal(scheduler_t* sched)
     //    stealing for generating far fewer block/unblock messages.
     if (!block_sent)
     {
-      if (steal_attempts < scheduler_count)
+      uint32_t current_active_scheduler_count = get_active_scheduler_count();
+      if (steal_attempts < current_active_scheduler_count)
       {
         steal_attempts++;
       }
-      else if ((!sched->asio_noisy) &&
-        ((tsc2 - tsc) > 1000000) &&
+      else if (((tsc2 - tsc) > 1000000) &&
         (ponyint_mutemap_size(&sched->mute_mapping) == 0))
       {
-        send_msg(0, SCHED_BLOCK, 0);
-        block_sent = true;
+        bool cmp_val = false;
+
+        // if we're the highest active scheduler thread
+        // and there are more active schedulers than the minimum requested
+        if ((sched == &scheduler[current_active_scheduler_count - 1])
+          && (current_active_scheduler_count > min_scheduler_count) &&
+          atomic_compare_exchange_strong_explicit(&scheduler_count_changing,
+          &cmp_val, true, memory_order_relaxed, memory_order_relaxed))
+
+        {
+          // let sched 0 know we're suspending
+          send_msg(0, SCHED_SUSPEND, 0);
+
+          // decrement active_scheduler_count so other schedulers know we're
+          // sleeping
+          atomic_fetch_sub_explicit(&active_scheduler_count, 1,
+            memory_order_relaxed);
+
+          cmp_val = true;
+
+          // unlock the bool that controls modifying the active scheduler count
+          // variable
+          atomic_compare_exchange_strong_explicit(&scheduler_count_changing,
```

Comment:
This can be a plain `atomic_store` since nobody will try to modify the variable while the current thread owns it.

The operation should have a release memory ordering since it marks the end of the critical section.

Reply:
Thanks for the suggestion. I've changed the logic accordingly.